internal/datastore/mongo_store.go

```diff
@@ -114,14 +144,87 @@ func (mongodb MongoDB) DeleteQueue(id db.QueueId) error {
 	return nil
 }
 
-func (mongodb MongoDB) AddTokenToQueue(db.QueueId, db.Token) {
-	panic("Not implemented")
+func (mongodb MongoDB) AddTokenToQueue(id db.QueueId, token db.Token) (db.TokenId, error) {
+	token.Id = ""
+	token.QueueId = id
+
+	// Lock queue to ensure 2 concurrent add tokens don't happen
+	queueMutex.Lock()
+
+	// if there were 2 concurrent calls to this function, 2 tokens might get the same number
+	max, err := mongodb.GetMaxToken(id)
+
+	if err != nil {
+		log.Fatal(err)
+		queueMutex.Unlock()
+		return token.Id, err
```

Comment:
Here we forget to release the lock. 

Is there some golang construct that would make sure the lock is always released when the function returns or throws?

Reply:
I think using `defer queueMutex.Unlock()` should unlock the mutex before a return. Right now I have manually done an unlock before every return statement. But I think using `defer` is better. I'll make the change.